warp/yul/VariableInliner.py

```diff
@@ -0,0 +1,80 @@
+from __future__ import annotations
+
+from typing import Dict, List, Optional, Tuple, Union
+
+import yul.yul_ast as ast
+from yul.AstMapper import AstMapper
+
+Scope = Dict[str, ast.Literal]
+
+
+class VariableInliner(AstMapper):
+    """This class inlines the value of variables by tracking their values
+    across declarations and assignments.
+    """
+
+    def __init__(self):
+        super().__init__()
+        self.scope: List[Scope] = list()
+
+    def visit_block(self, node: ast.Block) -> ast.Block:
+        self.scope.append(dict())
+        visited_node = ast.Block(tuple(self.visit_list(node.statements)))
+        self.scope.pop()
+        return visited_node
+
+    def visit_variable_declaration(
+        self, node: ast.VariableDeclaration
+    ) -> ast.VariableDeclaration:
+        current_scope = self.scope[-1]
+        # Since the rhs is either a literal or an identifier only one variable
```

Comment:
We should also mention as much to Julian

Reply:
Its only in the case where the rhs is `None` or its is a function call that you can have multiple variables, when its None then all of them are initialized to 0. This came up during my discussions with Julian but it just slipped from my mind when implementing stuff

Reply:
It's a little confusing here but the following lines do the checks for whether the RHS is literal or identifier. Function calls can't be inlined because of side-effects.

Reply:
I have updated the comment

Reply:
They don't check if there's even one variable. I moved the visitor in the pipeline here and there and it throws list index access exception in some cases

Reply:
I didn't realise that there can be a declaration without even one variable, seems like a bug?

Reply:
Well, it's just like in cairo `let () = add()`. Is it necessarily a bug? Perhaps, I'll need to check more carefully. You can reproduce this by moving the three visitors to the end of the pipeline

Reply:
Its definitely not valid Yul, but I have added the check.